src/main/java/edu/harvard/h2ms/service/report/ReportServiceImpl.java

```diff
@@ -0,0 +1,52 @@
+package edu.harvard.h2ms.service.report;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.PostConstruct;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import org.springframework.stereotype.Service;
+
+/** Loads and caches all report workers, as well as retrieve workers using report type names */
+@Component
+@Service("reportService")
+public class ReportServiceImpl implements ReportService {
+
+  final Logger log = LoggerFactory.getLogger(ReportServiceImpl.class);
+
+  /* this is where all the reportWorkers are loaded */
+  @Autowired private List<ReportWorker> reportWorkers;
+
+  private static final Map<String, ReportWorker> reportWorkerCache = new HashMap<>();
+
+  /** Initialize report worker cache */
+  @PostConstruct
+  public void initReportWorkerCache() {
+    for (ReportWorker reportWorker : reportWorkers) {
+      log.info("*********** cache init on" + reportWorker.getType());
+      reportWorkerCache.put(reportWorker.getType(), reportWorker);
+    }
+  }
+
+  public static ReportWorker getReportWorker(String reportType) {
+    ReportWorker reportWorker = reportWorkerCache.get(reportType);
+    if (reportWorker == null) throw new RuntimeException(reportType);

```

Comment:
For some reason I'm getting an error on the notification:

2018-04-28 13:07:33.154  INFO 1076 --- [pool-2-thread-1] e.h.h.s.report.NotificationServiceImpl   : user admin@h2ms.org is ready to be notified
2018-04-28 13:07:33.154  INFO 1076 --- [pool-2-thread-1] e.h.h.service.report.ReportServiceImpl   : report requested.  report type: complianceWarningMonthly
2018-04-28 13:07:33.156 ERROR 1076 --- [pool-2-thread-1] o.s.s.s.TaskUtils$LoggingErrorHandler    : Unexpected error occurred in scheduled task.

java.lang.RuntimeException: complianceWarningMonthly
	at edu.harvard.h2ms.service.report.ReportServiceImpl.getReportWorker(ReportServiceImpl.java:36) ~[classes!/:0.0.1-SNAPSHOT]
	at edu.harvard.h2ms.service.report.ReportServiceImpl.requestReport(ReportServiceImpl.java:45) ~[classes!/:0.0.1-SNAPSHOT]
	at edu.harvard.h2ms.service.report.NotificationServiceImpl.notifyUsers(NotificationServiceImpl.java:62) ~[classes!/:0.0.1-SNAPSHOT]
	at edu.harvard.h2ms.service.report.NotificationServiceImpl.pollNotifications(NotificationServiceImpl.java:37) ~[classes!/:0.0.1-SNAPSHOT]
	at sun.reflect.GeneratedMethodAccessor241.invoke(Unknown Source) ~[na:na]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_161]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_161]
	at org.springframework.scheduling.support.ScheduledMethodRunnable.run(ScheduledMethodRunnable.java:65) ~[spring-context-4.3.14.RELEASE.jar!/:4.3.14.RELEASE]
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54) ~[spring-context-4.3.14.RELEASE.jar!/:4.3.14.RELEASE]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_161]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [na:1.8.0_161]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [na:1.8.0_161]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [na:1.8.0_161]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_161]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_161]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_161]


Do you know why it can't find the notification?

Reply:
Forgot to add @Component.  Added.